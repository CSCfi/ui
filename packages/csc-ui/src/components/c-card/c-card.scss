:host {
  --c-card-gap: 24px;

  border-radius: 6px;
  box-shadow: rgb(0 0 0 / 15%) 0px 10px 20px, rgb(0 0 0 / 20%) 0px 5px 5px;
  display: flex;
  flex-direction: column;
  gap: var(--c-card-gap);
  padding: var(--c-card-gap) 0;
  position: relative;
}

::slotted(c-loader) {
  border-radius: 6px;
}
