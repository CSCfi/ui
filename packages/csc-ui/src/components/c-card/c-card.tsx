import { Component, Element, Host, h, Prop, getAssetPath } from '@stencil/core';

export type CardBackground = 'puhti' | 'mahti' | 'allas';
/**
 * @group Cards
 * @slot - Card contents
 */
@Component({
  tag: 'c-card',
  styleUrl: 'c-card.scss',
  shadow: true,
  assetsDirs: ['assets'],
})
export class CCard {
  /**
   * Card background image for login pages of specific services
   */
  @Prop() background: CardBackground;

  /**
   * Background color
   */
  @Prop() backgroundColor: string = 'white';

  @Element() host: HTMLCCardElement;

  private _allowedBackgrounds = ['puhti', 'mahti', 'allas'];

  render() {
    const style = {
      'background-color': this.backgroundColor,
    };

    const hostClasses = {
      'c-card': true,
    };

    if (this._allowedBackgrounds.includes(this.background)) {
      style['background-image'] = `url(${getAssetPath(
        `./assets/${this.background}.gif`,
      )}`;
      style['background-size'] = 'cover';
      style['background-position-y'] = 'bottom';
    }

    return (
      <Host class={hostClasses} style={style}>
        <slot></slot>
      </Host>
    );
  }
}
