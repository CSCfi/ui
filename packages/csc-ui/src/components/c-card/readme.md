# c-card



<!-- Auto Generated Below -->


## Properties

| Property     | Attribute    | Description                         | Type      | Default |
| ------------ | ------------ | ----------------------------------- | --------- | ------- |
| `fullscreen` | `fullscreen` | Enable the fullscreen toggle button | `boolean` | `false` |


## Methods

### `enterFullscreen() => Promise<void>`

Enter fullscreen from the outside

#### Returns

Type: `Promise<void>`



### `exitFullscreen() => Promise<void>`

Exit fullscreen from the outside

#### Returns

Type: `Promise<void>`




## Slots

| Slot             | Description     |
| ---------------- | --------------- |
| `"Default slot"` | Card components |


## CSS Custom Properties

| Name                        | Description               |
| --------------------------- | ------------------------- |
| `--c-card-background-color` | Card background color     |
| `--c-card-gap`              | Card column gap / padding |


## Dependencies

### Depends on

- [c-icon-button](../c-icon-button)
- [c-icon](../c-icon)

### Graph
```mermaid
graph TD;
  c-card --> c-icon-button
  c-card --> c-icon
  c-icon-button --> c-badge
  c-icon-button --> c-ripple
  style c-card fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
