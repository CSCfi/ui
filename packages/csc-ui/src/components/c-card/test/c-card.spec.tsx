import { newSpecPage } from '@stencil/core/testing';
import { CCard } from '../c-card';

describe('c-card', () => {
  it('renders default html', async () => {
    const page = await newSpecPage({
      components: [CCard],
      html: `<c-card></c-card>`,
    });

    expect(page.root).toMatchSnapshot();
  });

  it('render default with text', async () => {
    const page = await newSpecPage({
      components: [CCard],
      html: `
      <c-card>
        <c-card-title>The title</c-card-title>
        <c-card-content>
          Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut
          labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco
          laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in
          voluptate velit esse cillum dolore eu fugiat nulla pariatur.
        </c-card-content>
      </c-card>`,
    });

    expect(page.root).toMatchSnapshot();
  });

  it('render custom background-color', async () => {
    const page = await newSpecPage({
      components: [CCard],
      html: `<c-card background-color="red"></c-card>`,
    });

    expect(page.root).toMatchSnapshot();
  });

  it('render chosen background-photo, called allas', async () => {
    const page = await newSpecPage({
      components: [CCard],
      html: `<c-card background="allas"></c-card>`,
    });

    expect(page.root).toMatchSnapshot();
  });

  it('render another chosen background-photo, called mahti', async () => {
    const page = await newSpecPage({
      components: [CCard],
      html: `<c-card background="mahti"></c-card>`,
    });

    expect(page.root).toMatchSnapshot();
  });

  it('render custom background-photo', async () => {
    const page = await newSpecPage({
      components: [CCard],
      html: `<c-card background="dog"></c-card>`,
    });

    expect(page.root).toMatchSnapshot();
  });

  it('with fullscreen button', async () => {
    const page = await newSpecPage({
      components: [CCard],
      html: `<c-card fullscreen={true}></c-card>`,
    });

    expect(page.root).toMatchSnapshot();
  });

  it('opens fullscreen when button pressed', async () => {
    const page = await newSpecPage({
      components: [CCard],
      html: `<c-card fullscreen={true}></c-card>`,
    });

    expect(page.root).toMatchSnapshot();

    const iconbutton = page.root.shadowRoot.querySelector(
      '.c-card__fullscreen-toggle',
    ) as HTMLElement;

    iconbutton.click();

    await page.waitForChanges();

    expect(page.root).toMatchSnapshot();
  });

  it('shrinks back to normal when button pressed in fullscreen-mode', async () => {
    const page = await newSpecPage({
      components: [CCard],
      html: `<c-card fullscreen={true}></c-card>`,
    });

    expect(page.root).toMatchSnapshot();

    const iconbutton = page.root.shadowRoot.querySelector(
      '.c-card__fullscreen-toggle',
    ) as HTMLElement;

    iconbutton.click();

    await page.waitForChanges();

    expect(page.root).toMatchSnapshot();

    iconbutton.click();

    await page.waitForChanges();

    expect(page.root).toMatchSnapshot();
  });
});
