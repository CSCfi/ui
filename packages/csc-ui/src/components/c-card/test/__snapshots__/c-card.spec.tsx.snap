// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`c-card opens fullscreen when button pressed 1`] = `
<c-card class="c-card" fullscreen="{true}" style="background-color: white;">
  <mock:shadow-root>
    <c-icon-button aria-hidden="true" class="c-card__fullscreen-toggle" size="small" text="" title="Enter fullscreen">
      <c-icon path="M5,5H10V7H7V10H5V5M14,5H19V10H17V7H14V5M17,14H19V19H14V17H17V14M10,17V19H5V14H7V17H10Z"></c-icon>
    </c-icon-button>
    <slot></slot>
  </mock:shadow-root>
</c-card>
`;

exports[`c-card opens fullscreen when button pressed 2`] = `
<c-card class="c-card c-card--fullscreen" fullscreen="{true}" style="background-color: white;">
  <mock:shadow-root>
    <c-icon-button aria-hidden="true" class="c-card__fullscreen-toggle" size="small" text="" title="Exit fullscreen">
      <c-icon path="M14,14H19V16H16V19H14V14M5,14H10V19H8V16H5V14M8,5H10V10H5V8H8V5M19,8V10H14V5H16V8H19Z"></c-icon>
    </c-icon-button>
    <slot></slot>
  </mock:shadow-root>
</c-card>
`;

exports[`c-card render another chosen background-photo, called mahti 1`] = `
<c-card background="mahti" class="c-card" style="background-color: white; background-image: url(/assets/mahti.gif; background-size: cover; background-position-y: bottom;">
  <mock:shadow-root>
    <slot></slot>
  </mock:shadow-root>
</c-card>
`;

exports[`c-card render chosen background-photo, called allas 1`] = `
<c-card background="allas" class="c-card" style="background-color: white; background-image: url(/assets/allas.gif; background-size: cover; background-position-y: bottom;">
  <mock:shadow-root>
    <slot></slot>
  </mock:shadow-root>
</c-card>
`;

exports[`c-card render custom background-color 1`] = `
<c-card background-color="red" class="c-card" style="background-color: red;">
  <mock:shadow-root>
    <slot></slot>
  </mock:shadow-root>
</c-card>
`;

exports[`c-card render custom background-photo 1`] = `
<c-card background="dog" class="c-card" style="background-color: white;">
  <mock:shadow-root>
    <slot></slot>
  </mock:shadow-root>
</c-card>
`;

exports[`c-card render default with text 1`] = `
<c-card class="c-card" style="background-color: white;">
  <mock:shadow-root>
    <slot></slot>
  </mock:shadow-root>
  <c-card-title>
    The title
  </c-card-title>
  <c-card-content>
    Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur.
  </c-card-content>
</c-card>
`;

exports[`c-card renders default html 1`] = `
<c-card class="c-card" style="background-color: white;">
  <mock:shadow-root>
    <slot></slot>
  </mock:shadow-root>
</c-card>
`;

exports[`c-card shrinks back to normal when button pressed in fullscreen-mode 1`] = `
<c-card class="c-card" fullscreen="{true}" style="background-color: white;">
  <mock:shadow-root>
    <c-icon-button aria-hidden="true" class="c-card__fullscreen-toggle" size="small" text="" title="Enter fullscreen">
      <c-icon path="M5,5H10V7H7V10H5V5M14,5H19V10H17V7H14V5M17,14H19V19H14V17H17V14M10,17V19H5V14H7V17H10Z"></c-icon>
    </c-icon-button>
    <slot></slot>
  </mock:shadow-root>
</c-card>
`;

exports[`c-card shrinks back to normal when button pressed in fullscreen-mode 2`] = `
<c-card class="c-card c-card--fullscreen" fullscreen="{true}" style="background-color: white;">
  <mock:shadow-root>
    <c-icon-button aria-hidden="true" class="c-card__fullscreen-toggle" size="small" text="" title="Exit fullscreen">
      <c-icon path="M14,14H19V16H16V19H14V14M5,14H10V19H8V16H5V14M8,5H10V10H5V8H8V5M19,8V10H14V5H16V8H19Z"></c-icon>
    </c-icon-button>
    <slot></slot>
  </mock:shadow-root>
</c-card>
`;

exports[`c-card shrinks back to normal when button pressed in fullscreen-mode 3`] = `
<c-card class="c-card" fullscreen="{true}" style="background-color: white;">
  <mock:shadow-root>
    <c-icon-button aria-hidden="true" class="c-card__fullscreen-toggle" size="small" text="" title="Enter fullscreen">
      <c-icon path="M5,5H10V7H7V10H5V5M14,5H19V10H17V7H14V5M17,14H19V19H14V17H17V14M10,17V19H5V14H7V17H10Z"></c-icon>
    </c-icon-button>
    <slot></slot>
  </mock:shadow-root>
</c-card>
`;

exports[`c-card with fullscreen button 1`] = `
<c-card class="c-card" fullscreen="{true}" style="background-color: white;">
  <mock:shadow-root>
    <c-icon-button aria-hidden="true" class="c-card__fullscreen-toggle" size="small" text="" title="Enter fullscreen">
      <c-icon path="M5,5H10V7H7V10H5V5M14,5H19V10H17V7H14V5M17,14H19V19H14V17H17V14M10,17V19H5V14H7V17H10Z"></c-icon>
    </c-icon-button>
    <slot></slot>
  </mock:shadow-root>
</c-card>
`;
